presto-main/src/main/java/com/facebook/presto/sql/planner/GroupingOperationRewriter.java

```diff
@@ -0,0 +1,197 @@
+import java.util.stream.Collectors;
+
+import static com.facebook.presto.operator.scalar.GroupingOperationFunction.GROUPING;
+import static com.facebook.presto.spi.type.BigintType.BIGINT;
+import static com.facebook.presto.spi.type.IntegerType.INTEGER;
+import static com.facebook.presto.sql.analyzer.ExpressionAnalyzer.resolveFunction;
+import static com.facebook.presto.util.ImmutableCollectors.toImmutableList;
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+public class GroupingOperationRewriter
+        extends ExpressionRewriter<Void>
+{
+    private final QuerySpecification queryNode;
+    private final Analysis analysis;
+    private final Metadata metadata;
+    private boolean containsGroupingOperation = false;
+
+    public GroupingOperationRewriter(QuerySpecification queryNode, Analysis analysis, Metadata metadata)
+    {
+        this.queryNode = requireNonNull(queryNode, "node is null");
+        this.analysis = requireNonNull(analysis, "analysis is null");
+        this.metadata = requireNonNull(metadata, "metadata is null");
+    }
+
+    @Override
+    public Expression rewriteExpression(Expression node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+    {
+        return treeRewriter.defaultRewrite(node, context);
+    }
+
+    @Override
+    public Expression rewriteFunctionCall(FunctionCall node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+    {
+        FunctionCall functionCall = treeRewriter.defaultRewrite(node, context);
+
+        if (functionCall != node && containsGroupingOperation) {
+            FunctionCall rewrittenFunction = functionCall;
+            if (functionCall.getWindow().isPresent()) {
+                List<Expression> rewrittenPartitionBy = rewrittenFunction.getWindow().get().getPartitionBy();
+                List<Expression> originalPartitionBy = node.getWindow().get().getPartitionBy();
+                checkState(
+                        originalPartitionBy.size() == rewrittenPartitionBy.size(),
+                        "the original partition by clause and the rewritten one are not equal in size"
+                );
+
+                for (int i = 0; i < rewrittenPartitionBy.size(); i++) {
+                    Type originalType = analysis.getType(originalPartitionBy.get(i));
+                    IdentityHashMap<Expression, Type> newTypes = new IdentityHashMap<>();
+                    newTypes.put(rewrittenPartitionBy.get(i), originalType);
+                    analysis.addTypes(newTypes);
+                }
+
+                List<SortItem> rewrittenOrderBy = rewrittenFunction.getWindow().get().getOrderBy();
+                List<SortItem> originalOrderBy = node.getWindow().get().getOrderBy();
+                checkState(
+                        originalOrderBy.size() == rewrittenOrderBy.size(),
+                        "the original order by clause and the rewritten one are not equal in size"
```

Comment:
I think the assumption for check state is that they're not user visible errors (just like requireNonNull)